### ISSUE TITLE
Concern about RNN based language model (ptb_reader.py)

### ISSUE THREAD

**Issue by rbaral** [tensorflow/models#2043]
I have a question regarding the ptb_reader module at [https://github.com/tensorflow/models/blob/master/tutorials/rnn/ptb/reader.py] that was the part of RNN based language modeling.

It has a method 
`def ptb_producer(raw_data, batch_size, num_steps, name=None):`

which should return the batch of data in the form of tensors. It attempted to do the following for this:
```python
i = tf.train.range_input_producer(epoch_size, shuffle=False).dequeue()
x = tf.strided_slice(data, [0, i * num_steps],
                     [batch_size, (i + 1) * num_steps])
x.set_shape([batch_size, num_steps])
y = tf.strided_slice(data, [0, i * num_steps + 1],
                     [batch_size, (i + 1) * num_steps + 1])
y.set_shape([batch_size, num_steps])
return x, y
```
It seems like it just takes very small subset of data rather than transforming whole data into small batches.

I am not sure if this is an issue or intended.

Also, separate models are created for the training, test and validation dataset and the perplexity is calculated separately. Why not to train the model with the training data and use the same model to predict on the test data to calculate the loss?

**Comment by reedwm** [comment]
This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. There is also a larger community that reads questions there. Thanks!

This isn't an issue because every time `x` and `y` are evaluated, `i` will be a different value and so a different batch of data will be retrieved. The first time `i` is evaluated, it will be 0, the second time it will be 1 and so on. I would recommend reading [Threading and Queueing](https://www.tensorflow.org/programmers_guide/threading_and_queues) to learn more, as `tf.train.range_input_producer` is implemented with a queue.

The models for training, test and validation are almost identical. The only difference is that in the test and validation models, dropout isn't used. Note that the test and validation models use the same parameters as the training model, since `reuse=not is_training` is passed to the `tf.contrib.rnn.BasicLSTMCell` constructor.
